File: app/src/main/java/ska/sudoku/SolverTask.java
package ska.sudoku;

import android.os.AsyncTask;

import java.util.Arrays;
import java.util.List;

class SolverTask extends AsyncTask<Cell, Void, SolverTask.Result> {

    interface SolverCallback {
        void onFinished(Result result);
    }

    private final int max;
    private int currentCell;
    private final SolverCallback listener;
    private static long startTime;
    private final Solver solver;

    SolverTask(int max, SolverCallback listener) {
        this.max = max;
        this.listener = listener;
        solver = new Solver(max);
    }

    @Override
    protected void onPreExecute() {
        super.onPreExecute();
        startTime = System.currentTimeMillis();
    }

    @Override
    protected Result doInBackground(Cell... params) {
        List<Cell> grid = Arrays.asList(params);
        if (!solver.isSolvable(grid)) {
            cancel(true);
            return new Result(grid, Result.Code.NOT_SOLVABLE);
        }

        boolean lastResult = true;
        while (currentCell < (max * max)) {
            if (isCancelled()) {
                return new Result(grid, Result.Code.USER_CANCELLED);
            }

            Cell cell = grid.get(currentCell);
            if (cell.isPreFilled()) {
                setCurrentCell(lastResult);
                continue;
            }
            lastResult = solver.solveCell(grid, cell);
            setCurrentCell(lastResult);
            Thread.yield();
        }
        return new Result(grid);
    }

    @Override
    protected void onPostExecute(Result result) {
        super.onPostExecute(result);
        System.out.println(result.getGrid().toString());
        if (listener != null)
            listener.onFinished(result);
    }

    @Override
    protected void onCancelled(Result result) {
        if (listener != null) listener.onFinished(result);
    }

    private void setCurrentCell(boolean success) {
        if (success)
            currentCell++;
        else
            currentCell--;
    }

    static class Result {
        enum Code { NO_ERROR, NOT_SOLVABLE, USER_CANCELLED }

        private final List<Cell> grid;
        private final Code error;
        private final long duration;

        Result(List<Cell> grid) {
            this(grid, Code.NO_ERROR);
        }

        Result(List<Cell> grid, Code error) {
            this.grid = grid;
            this.error = error;
            this.duration = calculateDuration();
        }

        List<Cell> getGrid() {
            return grid;
        }

        long getDuration() {
            return duration;
        }

        Code getError() {
            return error;
        }

        private long calculateDuration() {
            return System.currentTimeMillis() - startTime;
        }
    }
}

File: app/src/main/java/ska/sudoku/SolverTask.kt
package ska.sudoku

import android.os.AsyncTask

class SolverTask(private val max: Int, private val callback: (Result) -> Unit) :
        AsyncTask<Cell, Void, Result>() {

    companion object {
        var startTime: Long = 0
    }

    private val solver: Solver by lazy {
        Solver(max)
    }
    private var currentCell: Int = 0

    override fun onPreExecute() {
        super.onPreExecute()
        startTime = System.currentTimeMillis()
    }

    override fun doInBackground(vararg params: Cell): Result {
        val grid: List<Cell> = params.toList()
        if (!solver.isSolvable(grid)) {
            cancel(true)
            return Result(grid, Result.Code.NOT_SOLVABLE)
        }

        var lastResult = true
        while (currentCell < (max * max)) {
            if (isCancelled) {
                return Result(grid, Result.Code.USER_CANCELLED)
            }

            val cell = grid[currentCell]
            if (cell.state == Cell.State.USER) {
                nextCell(lastResult)
                continue
            }
            lastResult = solver.solveCell(grid, cell)
            nextCell(lastResult)
            Thread.yield()
        }
        grid.filter { it.state != Cell.State.USER }
                .forEach { it.markAsSolved() }
        return Result(grid)
    }

    override fun onPostExecute(result: Result) {
        super.onPostExecute(result)
        System.out.println(result.grid.toString())
        callback.invoke(result)
    }

    override fun onCancelled(result: Result) {
        super.onCancelled(result)
        callback.invoke(result)
    }

    private fun nextCell(success: Boolean) {
        if (success) currentCell++
        else currentCell--
    }
}
